lib/recursive-methods.rb

```diff
@@ -1,49 +1,119 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    if n < 0 
+        raise ArgumentError,'only input positivie integer'
+    end 
+    if (n == 1 || n == 0 )
+        return 1
+    else 
+        return n * factorial(n-1)
+    end 
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: 0(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    if s.length <=1 
+        return s 
+    end 
+    reverseword = ""
+    i = s.length 
+    while i > 0 
+        reverseword += s[i-1]
+        i -= 1
+    end 
+    return reverseword
+
 end
 
-# Time complexity: ?
-# Space complexity: ?
+
+
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
+    if s.length <= 1
+        return s 
+    else  
+        return s[-1] + reverse(s[0..-2]) 
+    end 
     raise NotImplementedError, "Method not implemented"
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
+    if n == 0
+        return 0
+    else  
+        return 2 + bunny(n-1)
+    end 
     raise NotImplementedError, "Method not implemented"
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def nested(s)
```

Comment:
Since you create a new string with each recursive call, the space complexity is O(n^2).